### PR TITLE
ENG-11392 let DR consumer complete transactions before shutting down

### PR DESCRIPTION
**Pull request by zeemanhe**
In graceful shutdown, check DR consumer stats. Let consumer apply all transactions before shutdown is issued.

### CODE REVIEW

File: lib/python/voltcli/checkstats.py [hunk 1]
```diff
@@ -208,3 +208,17 @@ def check_importer(runner):
...
         time.sleep(1)
+
+def check_dr_consumer(runner):
+     while True:
+        resp = get_stats(runner, 'DRCONSUMER')
+        outstanding = 0
+        if len(resp.table(1).tuples()) == 0:
+            return
+        for r in resp.table(1).tuples():
+            if r[7] <> r[8]:
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: Can you add comments here, why we need to check these two columns to determine DRConsumer is drained?

File: lib/python/voltcli/checkstats.py [hunk 2]
```diff
@@ -208,3 +208,17 @@ def check_importer(runner):
...
+                outstanding += 1
+                runner.info('Partition %d on host %d has outstanding DR consumer transactions.' %(r[4], r[1]))
+        if outstanding == 0:
+            return
+        time.sleep(2)
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: Why sleep(2) here and other places sleep(1)?

File: lib/python/voltcli/voltadmin.d/shutdown.py [hunk 3]
```diff
@@ -44,12 +44,14 @@ def shutdown(runner):
             status = runner.call_proc('@Quiesce', [], []).table(0).tuple(0).column_integer(0)
             if status <> 0:
                 runner.abort('The cluster has failed to be quiesce with status: %d' % status)
-            runner.info('Completing outstanding export and DR transactions...')
+            runner.info('Completing outstanding export and DR producer transactions...')
```
Comment (xinjiacs) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: Maybe it's better to separate the exporter and dr producer check?
Let's discuss with Anish/Ning about the checking order here. I am not sure whether they matter or not,

File: lib/python/voltcli/voltadmin.d/shutdown.py [hunk 4]
```diff
@@ -44,12 +44,14 @@ def shutdown(runner):
...
             checkstats.check_export_dr(runner)
             runner.info('Completing outstanding client transactions.')
             checkstats.check_clients(runner)
             runner.info('Completing outstanding importer requests.')
             checkstats.check_importer(runner)
+            runner.info('Completing outstanding DR consumer transactions...')
+            checkstats.check_dr_consumer(runner)
```
Comment (xinjiacs) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: Also, does the --pause need to take care of these stats checking? It already has exporter and DR producer checked.

Comment (xinjiacs) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: How do you test it? Run the failed tests? or Manually?

Probably we can commit this and run the failed tests over and over on master.

Comment (zeemanhe) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: I have manually tested it to make sure the consumer stats are clear before shutdown is issued.  We can also do system testing on the branch.

We can certainly separate export and dr producer stats.

File: lib/python/voltcli/checkstats.py (hunk 1 above)
Comment (zeemanhe) [inline lib/python/voltcli/checkstats.py]: comments added

**Pull request closed without merging**